File: questModule/src/main/java/ru/nekit/android/qls/quest/mediator/textCamouflage/TextCamouflageContentMediator.java
package ru.nekit.android.qls.quest.mediator.textCamouflage;


import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Point;
import android.support.annotation.NonNull;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.devspark.robototextview.widget.RobotoTextView;

import java.util.ArrayList;
import java.util.List;

import ru.nekit.android.qls.R;
import ru.nekit.android.qls.quest.QuestContext;
import ru.nekit.android.qls.quest.mediator.shared.content.AbstractQuestContentMediator;
import ru.nekit.android.qls.quest.types.quest.TextQuest;
import ru.nekit.android.qls.utils.MathUtils;

public class TextCamouflageContentMediator extends AbstractQuestContentMediator {

    private static final Position START_POSITION = Position.TL;

    private List<String> dataForGrid;
    private List<TextView> mTextViews;

    private TextCamouflageContainer mTextCamouflageContainer;

    @Override
    public EditText getAnswerInput() {
        return null;
    }

    @Override
    public boolean includeInLayout() {
        return true;
    }

    @Override
    public void onCreateQuest(@NonNull QuestContext questContext, @NonNull ViewGroup rootContentContainer) {
        super.onCreateQuest(questContext, rootContentContainer);
        //mViewHolder = new TextCamouflageQuestViewHolder(questContext);
        TextQuest quest = (TextQuest) mQuest;
        // mViewHolder.textViewGrid.setHasFixedSize(true);
        mTextCamouflageContainer = new TextCamouflageContainer(mQuestContext);
        // ((ViewGroup) mViewHolder.getView()).addView(mTextCamouflageContainer);
        int additionSize = 0;
        while (additionSize == 0) {
            additionSize = MathUtils.randInt(-1, quest.questionStringArray[0].length());
        }
        RecyclerView.LayoutManager gridLayoutManager = new GridLayoutManager(mQuestContext,
                quest.questionStringArray[0].length() + additionSize,
                LinearLayoutManager.VERTICAL, false);
        // mViewHolder.textViewGrid.setLayoutManager(gridLayoutManager);
        dataForGrid = new ArrayList<>();
        mTextViews = new ArrayList<>();
        for (String questStringArrayItem : quest.questionStringArray) {
            final char[] charArray = questStringArrayItem.toCharArray();
            for (char aChar : charArray) {
                dataForGrid.add(String.valueOf(aChar).toUpperCase());
                RobotoTextView textView = new RobotoTextView(mQuestContext);
                textView.setText(String.valueOf(aChar).toUpperCase());
                textView.setBackground(mQuestContext.getResources().getDrawable(R.drawable.background_text_camouflage_label));
                textView.setTextColor(mQuestContext.getResources().getColor(R.color.black));
                textView.setGravity(Gravity.CENTER);
                mTextViews.add(textView);
                mTextCamouflageContainer.addView(textView);
            }
        }
        //  mViewHolder.textViewGrid.setAdapter(new TextCamouflageAdapter(dataForGrid));
    }

    @Override
    public void detachView() {
        super.detachView();
        //  mViewHolder.textViewGrid.setLayoutManager(null);
        //  mViewHolder.textViewGrid.setAdapter(null);
    }

    @Override
    public View getView() {
        return mTextCamouflageContainer;
    }

    @Override
    public void updateSize() {
        int width = mRootContentContainer.getWidth();
        int height = mRootContentContainer.getHeight();
        int savedHeight = height;
        int savedWidth = width;
        height = Math.min(width, height);
        Position position = START_POSITION;
        int positionIndex = Position.getIndex(position),
                stepX = 0,
                stepY = 0,
                bounceLeft = 0, bounceBottom = 0;
        boolean isVerticalDirection = (START_POSITION == Position.TL || START_POSITION == Position.BR);
        int textViewWidth = width / mTextViews.size() * 2,
                textViewHeight = height / mTextViews.size() * 2,
                textViewSize = Math.min(textViewWidth, textViewHeight) - 4;
        List<Point> pointList = new ArrayList<>();
        for (int i = 0; i < mTextViews.size(); i++) {
            TextView textView = mTextViews.get(i);
            if (i != 0) {
                positionIndex = (positionIndex + 1) % Position.values().length == 0 ? 0
                        : positionIndex + 1;
                position = Position.values()[positionIndex];
            }
            if (i != 0) {
                if (START_POSITION == position) {
                    if (isVerticalDirection) {
                        stepY++;
                    } else {
                        stepX++;
                    }
                }
            }
            int x = (position == Position.TL || position == Position.BL) ?
                    textViewWidth * stepX : width - textViewWidth * (stepX + 1);
            int y = (position == Position.TL || position == Position.TR) ?
                    textViewHeight * stepY : height - textViewHeight * (stepY + 1);
            Point point = new Point(x + textViewSize / 2, y + textViewSize / 2);
            pointList.add(point);
            bounceLeft = Math.max(bounceLeft, x + textViewSize);
            bounceBottom = Math.max(bounceBottom, y + textViewSize);
            textView.setY(y);
            textView.setX(x);
            if (i != 0) {
                if (START_POSITION == position) {
                    if (isVerticalDirection) {
                        stepX++;
                    } else {
                        stepY++;
                    }
                }
            }
            textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, textViewSize / 3);
            ViewGroup.LayoutParams textViewLayoutParams = textView.getLayoutParams();
            textViewLayoutParams.width = textViewSize;
            textViewLayoutParams.height = textViewSize;
        }
        mTextCamouflageContainer.setPointList(pointList);
        mTextCamouflageContainer.setX((savedWidth - bounceLeft) / 2);
        mTextCamouflageContainer.setY((savedHeight - bounceBottom) / 2);
        mTextCamouflageContainer.getLayoutParams().width = width;
        mTextCamouflageContainer.getLayoutParams().height = height;
    }

    private enum Position {
        TL,
        TR,
        BR,
        BL;

        public static int getIndex(Position position) {
            Position[] values = values();
            int index = 0;
            for (Position positionItem : values) {
                if (positionItem == position) {
                    return index;
                }
                index++;
            }
            return index;
        }
    }

    enum VisualType {
        SPIRAL,
        SIN
    }

    private class TextCamouflageContainer extends FrameLayout {

        private Paint paint;
        private List<Point> pointList;

        public TextCamouflageContainer(@NonNull Context context) {
            super(context);
            setWillNotDraw(false);
            paint = new Paint();
            paint.setColor(Color.BLACK);
            paint.setStrokeWidth(10);
            paint.setStyle(Paint.Style.FILL_AND_STROKE);
            paint.setStrokeJoin(Paint.Join.ROUND);
        }

        @Override
        protected void onDetachedFromWindow() {
            paint = null;
            pointList.clear();
            pointList = null;
            super.onDetachedFromWindow();
        }

        public void setPointList(List<Point> pointList) {
            this.pointList = pointList;
        }

        @Override
        protected void onDraw(Canvas canvas) {
            super.onDraw(canvas);
            int count = getChildCount();
            for (int i = 0; i < count; i++) {
                if (i > 0) {
                    canvas.drawLine(pointList.get(i - 1).x, pointList.get(i - 1).y,
                            pointList.get(i).x,
                            pointList.get(i).y,
                            paint
                    );
                }
            }
        }
    }
}

File: questPresentation/src/main/java/ru/nekit/android/qls/quest/view/mediator/types/textCamouflage/TextCamouflageContentMediator.kt
package ru.nekit.android.qls.quest.view.mediator.types.textCamouflage

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Point
import android.support.v4.content.ContextCompat
import android.support.v7.widget.AppCompatTextView
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.TextView
import ru.nekit.android.qls.R
import ru.nekit.android.qls.domain.model.quest.Quest
import ru.nekit.android.qls.quest.QuestContext
import ru.nekit.android.qls.quest.types.TextQuest
import ru.nekit.android.qls.quest.view.mediator.content.SimpleContentMediator
import ru.nekit.android.utils.MathUtils
import java.util.*

//ver 1.0
class TextCamouflageContentMediator : SimpleContentMediator() {

    private var dataForGrid: MutableList<String>? = null
    private var textViews: MutableList<TextView>? = null

    private lateinit var _quest: TextQuest

    private var textCamouflageContainer: TextCamouflageContainer? = null

    override val answerInput: EditText?
        get() = null

    override val view: View?
        get() = textCamouflageContainer

    override fun includeInLayout(): Boolean {
        return true
    }

    override fun onCreate(questContext: QuestContext, quest: Quest) {
        super.onCreate(questContext, quest)
        _quest = quest as TextQuest
        //viewHolder = new TextCamouflageQuestViewHolder(questContext);
        //viewHolder.textViewGrid.setHasFixedSize(true);
        textCamouflageContainer = TextCamouflageContainer(questContext)
        //((ViewGroup) viewHolder.getView()).addView(textCamouflageContainer);
        var additionSize = 0
        while (additionSize == 0) {
            additionSize = MathUtils.randInt(-1, _quest.questionStringArray!![0].length)
        }
        val gridLayoutManager = GridLayoutManager(questContext,
                _quest.questionStringArray!![0].length + additionSize,
                LinearLayoutManager.VERTICAL, false)
        //viewHolder.textViewGrid.setLayoutManager(gridLayoutManager);
        dataForGrid = ArrayList()
        textViews = ArrayList()
        for (questStringArrayItem in _quest.questionStringArray!!) {
            val charArray = questStringArrayItem.toCharArray()
            for (aChar in charArray) {
                dataForGrid!!.add(aChar.toString().toUpperCase())
                val textView = AppCompatTextView(questContext)
                textView.text = aChar.toString().toUpperCase()
                textView.background = ContextCompat.getDrawable(questContext,
                        R.drawable.background_text_camouflage_label)
                textView.setTextColor(ContextCompat.getColor(questContext, R.color.black))
                textView.gravity = Gravity.CENTER
                textViews!!.add(textView)
                textCamouflageContainer!!.addView(textView)
            }
        }

        //  viewHolder.textViewGrid.setAdapter(new TextCamouflageAdapter(dataForGrid));
    }

    override fun detachView() {
        super.detachView()
        //  viewHolder.textViewGrid.setLayoutManager(null);
        //  viewHolder.textViewGrid.setAdapter(null);
    }

    override fun updateSize() {
        val width = rootContentContainer.width
        var height = rootContentContainer.height
        val savedHeight = height
        height = Math.min(width, height)
        var position = START_POSITION
        var positionIndex = Position.getIndex(position)
        var stepX = 0
        var stepY = 0
        var bounceLeft = 0
        var bounceBottom = 0
        val isVerticalDirection = START_POSITION == Position.TL || START_POSITION == Position.BR
        val textViewWidth = width / textViews!!.size * 2
        val textViewHeight = height / textViews!!.size * 2
        val textViewSize = Math.min(textViewWidth, textViewHeight) - 4
        val pointList = ArrayList<Point>()
        for (i in textViews!!.indices) {
            val textView = textViews!![i]
            if (i != 0) {
                positionIndex = if ((positionIndex + 1) % Position.values().size == 0)
                    0
                else
                    positionIndex + 1
                position = Position.values()[positionIndex]
            }
            if (i != 0) {
                if (START_POSITION == position) {
                    if (isVerticalDirection) {
                        stepY++
                    } else {
                        stepX++
                    }
                }
            }
            val x = if (position == Position.TL || position == Position.BL)
                textViewWidth * stepX
            else
                width - textViewWidth * (stepX + 1)
            val y = if (position == Position.TL || position == Position.TR)
                textViewHeight * stepY
            else
                height - textViewHeight * (stepY + 1)
            val point = Point(x + textViewSize / 2, y + textViewSize / 2)
            pointList.add(point)
            bounceLeft = Math.max(bounceLeft, x + textViewSize)
            bounceBottom = Math.max(bounceBottom, y + textViewSize)
            textView.y = y.toFloat()
            textView.x = x.toFloat()
            if (i != 0) {
                if (START_POSITION == position) {
                    if (isVerticalDirection) {
                        stepX++
                    } else {
                        stepY++
                    }
                }
            }
            textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, (textViewSize / 3).toFloat())
            val textViewLayoutParams = textView.layoutParams
            textViewLayoutParams.width = textViewSize
            textViewLayoutParams.height = textViewSize
        }
        textCamouflageContainer!!.setPointList(pointList)
        textCamouflageContainer!!.x = ((width - bounceLeft) / 2).toFloat()
        textCamouflageContainer!!.y = ((savedHeight - bounceBottom) / 2).toFloat()
        textCamouflageContainer!!.layoutParams.width = width
        textCamouflageContainer!!.layoutParams.height = height
    }

    private enum class Position {
        TL,
        TR,
        BR,
        BL;

        companion object {

            fun getIndex(position: Position): Int {
                val values = values()
                var index = 0
                for (positionItem in values) {
                    if (positionItem == position) {
                        return index
                    }
                    index++
                }
                return index
            }
        }
    }

    internal enum class VisualType {
        SPIRAL,
        SIN
    }

    private inner class TextCamouflageContainer(context: Context) : FrameLayout(context) {

        private lateinit var paint: Paint
        private lateinit var pointList: MutableList<Point>

        init {
            setWillNotDraw(false)
            paint = Paint()
            paint.color = Color.BLACK
            paint.strokeWidth = 10f
            paint.style = Paint.Style.FILL_AND_STROKE
            paint.strokeJoin = Paint.Join.ROUND
        }

        override fun onDetachedFromWindow() {
            pointList.clear()
            super.onDetachedFromWindow()
        }

        fun setPointList(pointList: MutableList<Point>) {
            this.pointList = pointList
        }

        override fun onDraw(canvas: Canvas) {
            super.onDraw(canvas)
            val count = childCount
            (0 until count)
                    .asSequence()
                    .filter { it > 0 }
                    .forEach {
                        canvas.drawLine(pointList[it - 1].x.toFloat(),
                                pointList[it - 1].y.toFloat(),
                                pointList[it].x.toFloat(),
                                pointList[it].y.toFloat(),
                                paint
                        )
                    }
        }
    }

    companion object {

        private val START_POSITION = Position.TL
    }
}
